### PR TITLE
Rewrite of CirclePokemon instance to spread Devices

### PR DESCRIPTION
**Pull request by Petap0w**
Attempt to modify the Device management in instances of type Circle Pokemon: instead of having each querying device assigned the next coord in instance list, all the assigned devices are spread evenly across the coord list.
The roundtime does not necessarily improve, but the devices being spread out gives a better coverage of nearby pokémons for IV Instances covering the area, raids and TR invasion detection.

Side changes:
- fixed typo in DeviceGroup page
- added ignored count on raws in Dashboard

Based on code from ControllerJS / VersX and modified to suit.

## Motivation and Context
On a large instance, leapfrog logic can leave an entire neighborhood unseen for quite a while.

## How Has This Been Tested?
Tested on base 1.20.3, assigning groups of 3 or 4 devices to different instances of 100+ coords each.
At startup the devices are randomly spread across the instance, and become after a while spread even.

## Screenshots (if appropriate):
**At start:**
Capture d’écran 2020-11-20 à 22 27 13
Capture d’écran 2020-11-20 à 22 27 22
**After 5 minutes:**
Capture d’écran 2020-11-20 à 22 32 19
Capture d’écran 2020-11-20 à 22 32 25
**After 20 more minutes:**
Capture d’écran 2020-11-20 à 22 53 37
Capture d’écran 2020-11-20 à 22 53 43

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by Fabio1988** [comment]
Can you fix swift-lint errors please?

**Comment by cks85** [comment]
> Can you fix swift-lint errors please?

I tried, but somehow there are now more errors than before on lines I didn't even change... i don't have much time to look for where to put the space correctly or which char should be upper or lower case to "look nice enough"...

**Comment by Fabio1988** [comment]
@cks85 it's perfectly described [here](https://github.com/RealDeviceMap/RealDeviceMap/pull/184/files). I would do it, but I have no rights :+1:

**Comment by 123FLO321** [comment]
If you need help with it let me know. 
Did you setup swiftlint locally?

**Comment by cks85** [comment]
> If you need help with it let me know.
> Did you setup swiftlint locally?

I'm not using xcode so can't really use all tools..
Last error is about length of function but i don't know how to shorten it more, already removed many things I don't even know if the logic will still work :D

File: Sources/RealDeviceMap/Controller/InstanceControllers/CircleInstanceController.swift [hunk 1]
```diff
@@ -41,35 +43,112 @@ class CircleInstanceController: InstanceControllerProto {
...
+        }
+        return ycoord + (coords.count - xcoord)
+    }
+
+    func checkSpacingDevices(uuid: String) -> [String: Int] {
+      let deadDeviceCutoffTime = Date().addingTimeInterval(-60)
+      var liveDevices = [String]()
+
+  		// Check if all registered devices are still online and clean list
+  		for (currentUuid, _) in currentUuidIndexes.sorted(by: { return $0.1 < $1.1 }) {
```
Comment (Fabio1988) [inline Sources/RealDeviceMap/Controller/InstanceControllers/CircleInstanceController.swift:62]: is it possible to autoformat this code? Should be available within IDE :+1:

**Comment by Pogodenhelder** [comment]
Running this PR now for a few hours.
Few things i noticed.

When a instance has only 1 device the roundtime seems a bit higher.. only a fraction
When a instance has multiple devices the roundtime seems a bit lower.. so it benefits a bit from this pr to lower the roundtime.
In the beginning roundtimes in the dashboard are way off, like it says 16 seconds while normally im around 600 seconds.
Keep running it normalize it to correct vallues.

Overal it looks like a better coverage in general.

**Review by DavisPoGo — DISMISSED** [review]
I've been running this PR for the past couple days. Based on the spread of devices, it is working as intended. I am not seeing a significant impact on round time but it improves raid reporting times.

**Comment by kbtbc** [comment]
I deployed this yesterday.  It seems to work, however the round trip times fluctuate so wildly and often show unrealistic times I am not confident in that data...   and without valid round trip data I am unable to conclude what, if any, benefit this PR offers. 

I am testing from 2 device instances to 20+ devices.  

In one simple test case, I had 5 devices on single routes of 20 locations.   Observed single device round trip times stayed in this range:  image

Combining those same 5 individual routes into one single route with the exact same devices/locations results in an instance that fluctuates between 0s (an impossible RT time) and over 500s:   This route has been running for about 4 hours, and just now I check a couple of times and here are some samples ever couple minutes:
63s
image
then 0s
image
and this will show very high numbers like 470s, etc. ... it never seems to normalize, even after several hours. 

Phsyically checking the map 'last scanned' date of gyms from the above instance shows several gyms that haven't been checked in over 20 minutes.

From my experience so far, this pr doesn't seem to provide valid round-trip times, nor timely updates to all gyms.

**Comment by 123FLO321** [comment]
@kbtbc Gyms last scanned timestamp can be inaccurate. 
If nothing changes on a gym it will not write to the DB (and with memcache not even read the DB) and the last scanned timestamp will stay the same.

**Comment by kbtbc** [comment]
Here is an example from the above route, last scanned stated is about 25 minutes ago.  The egg hatched over 5 minutes ago, and still has not updated:
image

**Comment by kbtbc** [comment]
To further illustrate the unreliable nature of round trip times with this PR, here are several different instances (that have been running for a few hours already) and so I took screenshots of their round-trip times an hour apart:

image
 
 Many of those times are not even possible.  I think the idea that this eventually 'normalizes' is incorrect.   
 
 Would be nice to get accurate metrics...

**Comment by Petap0w** [comment]
Roundtime isn’t a good metric because with this PR the devices aren’t doing a round route. 
I believe this was already discussed on Discord. 
the goal isn’t to improve roundtime nor give more scan/hour but a wider, more even geographical coverage on instances with a “long route” spread widely on an area, and with few devices assigned to it. 
With the leapfrog behavior that is without this PR, you can imagine all your devices like a little train on one side of the city jumping over each other along the route going from one location to the next. So the other side of the city would never been scanned (read widely scanned for raids or nearbies or TR), until your train of devices have done half the route. 
With this PR each device jumps at a calculated location that is evenly distant from another device assigned on the route. That way all the area is seen evenly. 
Of course with routes of 20 locations per devices in average you’re better sticking to the normal behavior.
If you want metrics to see and compare scan rate, just check your iv scanned per hour on the area (db query, stats dashboard, or even set an IV type instance on a geofence covering the entire area).

**Comment by Fabio1988** [comment]
So everything is now also in #204 ? But cleaner ?

**Comment by Petap0w** [comment]
Replaced by PR #204

**Pull request closed without merging**